### PR TITLE
Discard transaction with invalid gas limit

### PR DESCRIPTION
**Pull request by 0xAleksaOpacic**
# Description

If a transaction has gasLimit bigger than the latest blockGasLimit it is considered invalid.
There is no way to know for sure what will be gas limit of the block where the transaction will be placed, so we have another check when we are creating a block.

Different error cases

Transaction has gasLimit > latestBlockGasLimit (it will not go into pool and it will be rejected as invalid)
Transaction has gasLimit < latestBlockGasLimit but when we create a block gasLimit > newBlockGasLimit (we will generate failed transaction receipt)
Testing:

Try to add a transaction with gasLimit 10000000001.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: txpool/mock_test.go [hunk 1]
```diff
@@ -13,7 +13,9 @@ type defaultMockStore struct {
 }
 
 func (m defaultMockStore) Header() *types.Header {
-	return &types.Header{}
+	return &types.Header{
+		GasLimit: 4712388,
```
Comment (0xAleksaOpacic) [inline txpool/mock_test.go:18]: @zivkovicmilos COMMENT FROM PREVIOUS PR THAT WE CLOSED
Is it possible to add a method to the defaultMockStore for setting the gas limit of the latest header?
Right now only 1 test uses this value I think, but in the future more unit tests may rely on this GasLimit value to be configurable

File: txpool/txpool.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ const (
 // errors
 var (
 	ErrIntrinsicGas        = errors.New("intrinsic gas too low")
+	ErrBlockLimitExceeded  = errors.New("transaction's gas limit exceeds block gas limit")
```
Comment (0xAleksaOpacic) [inline txpool/txpool.go]: Please make sure the error corresponds to the same one GETH uses:
https://github.com/ethereum/go-ethereum/blob/f32feeb260054d7c90c1efd57dfec6b717717e18/core/error.go#L60

This is required since many web3 frameworks hardcode these expected values in source code

File: e2e/txpool_test.go [hunk 3]
```diff
@@ -408,44 +408,6 @@ func TestTxPool_StressAddition(t *testing.T) {
 	}
 }
 
-func TestInvalidTransactionRecover(t *testing.T) {
```
Comment (0xAleksaOpacic) [inline e2e/txpool_test.go:411]: Please link the unit test(s) that cover this functionality in the new TxPool implementation 🙏

File: txpool/mock_test.go (hunk 1 above)
Comment (0xAleksaOpacic) [reply txpool/mock_test.go:18]: Not directly, because defaultMockStore is mapped to store object, and we can call only mock methods from that interface. There are a few ways we can bypass this. One idea comes to my mind.
1. create a constructor for initialization defaultMockStore object. Which can receive header as an attribute
2. when initializing the pool, we can send header data with what we want to initialize

But this will complicate pool initialization. Will research for better solution

File: e2e/txpool_test.go (hunk 3 above)
Comment (0xAleksaOpacic) [reply e2e/txpool_test.go:411]: This PR will discard every transaction if they are sent with an invalid gas limit, no transaction will pass validationTx with an invalid gas limit. 
There is no way to write unit/e2e test which will test edge-case when transaction is valid when arrives in txPool (latest block is valid) but new block is invalid (while building)

File: txpool/txpool.go (hunk 2 above)
Comment (0xAleksaOpacic) [reply txpool/txpool.go]: Fixed c74ebfd5ba1cd4b095ad14db50d864d7101d459f

File: e2e/txpool_test.go (hunk 3 above)
Comment (zivkovicmilos) [reply e2e/txpool_test.go:411]: Why not test adding the transaction?

You can have a few test cases for valid gas limits, and some for invalid

File: txpool/mock_test.go (hunk 1 above)
Comment (zivkovicmilos) [reply txpool/mock_test.go:18]: Can we go with solution 1?

I don't feel good about hardcoding these values, even if it's a mock store. Someone will need to untangle it in some other PR if we leave it like this

**Review by zivkovicmilos — APPROVED** [review]
Approving again 👍 

I've left some comments for discussion, otherwise it looks good 💯

**Review by brkomir — APPROVED** [review]
LGTM

File: txpool/mock_test.go (hunk 1 above)
Comment (0xAleksaOpacic) [reply txpool/mock_test.go:18]: This is one possible solution. It is a bit redundant with sending nil value where we want a default value. But go does not like optional parameters. 55eeead6ab4c2431394e0611e1e6969516e6f74f

File: e2e/txpool_test.go (hunk 3 above)
Comment (0xAleksaOpacic) [reply e2e/txpool_test.go:411]: In the txpool_test.go we have a unit test that covers adding error tx (TestAddTxErrors). It covers all validation rules and because I have added a new validation rule I had to add a new test case for this rule. 

https://github.com/0xPolygon/polygon-edge/pull/349/files#diff-c6d10affe0540289b09ab01da147d827eb2c3aa0a899b5e5d5c43810821422cbR113

In my opinion, because this is mock with defined values this one test covers the validation check for ErrBlockLimitExceeded. Adding multiple test cases for valid and invalid values would just be redundant check if one value is bigger than the other.

And every other test which adds a new transaction inside this package will show that if we have a valid gas limit (tx.gasLimit < latestBlockGasLimit) will be considered valid

**Review by Kourin1996 — APPROVED** [review]
Looks good

**Review by dbrajovic — APPROVED** [review]

File: txpool/txpool_test.go [hunk 4]
```diff
@@ -69,11 +70,11 @@ func newTx(addr types.Address, nonce, slots uint64) *types.Transaction {
 }
 
 // returns a new txpool with default test config
-func newTestPool() (*TxPool, error) {
+func newTestPool(header *types.Header) (*TxPool, error) {
```
Comment (brkomir) [inline txpool/txpool_test.go:73]: Why this parameter? Not used anywhere, you pass "nil" for every func call...

Comment (lazartravica) [reply txpool/txpool_test.go:73]: @Aleksao998 please address this comment

Comment (0xAleksaOpacic) [reply txpool/txpool_test.go:73]: @zivkovicmilos requested a change to make that we can adjust header when are creating the pool in our tests and not to have default values. I agree that this is not a pretty solution at all. But was the only one I have found. I will tag both of you in this discussion.

File: txpool/mock_test.go (hunk 1 above)
Comment (0xAleksaOpacic) [reply txpool/mock_test.go:18]: @brkomir @lazartravica Here is the discussion why i had to add header parameter

File: txpool/txpool_test.go (hunk 4 above)
Comment (zivkovicmilos) [reply txpool/txpool_test.go:73]: What about adding a wrapper mock struct that embeds TxPool and has a method `SetHeader` or something like that?

Comment (0xAleksaOpacic) [reply txpool/txpool_test.go:73]: I am not really sure what do you think
Store interface does not have method setHeader or something like that. So what we pass to the constructor when we are creating the store. These will be value when we get them.
